### PR TITLE
Add basic authentication

### PR DESCRIPTION
**Pull request by brgeen**
**Update Client Credentials flow for Courses API** [Asana Task](https://app.asana.com/0/1206395725449160/1206174762676768/f)

### Changes

This adds logic for the [Password Grant](https://www.oauth.com/oauth2-servers/access-tokens/password-grant/) flow. Which allows us to send Basic Auth (username/password) credentials to Accounts Next. 

### Test
Reference [Courses API PR](https://github.com/Arux-Software/onsite_courses_api/pull/37)

### CODE REVIEW

**Review by nhaglind — APPROVED** [review]

File: lib/arux_app/api/auth.rb [hunk 1]
```diff
@@ -60,6 +60,32 @@ def authorization_url(scope: "public")
...
+          scope: scope,
+          grant_type: "password",
+          client_id: client_id,
+          client_secret: client_secret
+        }
+        request = HTTPI::Request.new
+        request.url = "#{self.class.server_uri}/oauth/token"
+        request.body = params
+        request.headers = { 'User-Agent' => USER_AGENT }
+        request.auth.basic(username, password)
```
Comment (goronfreeman) [inline lib/arux_app/api/auth.rb]: I looked at the [source](https://github.com/savonrb/httpi/blob/main/lib/httpi/request.rb#L17-L23) for `HTTPI::Request`, and it looks like you can do this:

```suggestion
        request = HTTPI::Request.new(
          url: "#{self.class.server_uri}/oauth/token",
          body: params,
          headers: { "User-Agent" => USER_AGENT }
        )
        request.auth.basic(username, password)
```

That still leaves `request.auth.basic` that you can't pass in. You could also do:

```rb
        request = HTTPI::Request.new.tap do |req|
          req.url = "#{self.class.server_uri}/oauth/token"
          req.body = params
          req.headers = { 'User-Agent' => USER_AGENT }
          req.auth.basic(username, password)
        end
```

This is a super minor thing, but I like to be able to visually see a chunk of stuff that is all working to build up an object as one big thing, not a series of separate lines, if that makes sense. Just makes it a little more readable for me personally.

File: lib/arux_app/api/auth.rb [hunk 2]
```diff
@@ -60,6 +60,32 @@ def authorization_url(scope: "public")
...
+
+        if !response.error?
+          AccessToken.new(
+            token: JSON.parse(response.body)['access_token'],
+            scope: JSON.parse(response.body)['scope'],
+            auth: self
+          )
+        else
+          raise(API::Error.new(response.code, response.body))
+        end
```
Comment (goronfreeman) [inline lib/arux_app/api/auth.rb]: You could do a guard here:

```suggestion
        raise(API::Error.new(response.code, response.body)) if response.error?
        
        AccessToken.new(
          token: JSON.parse(response.body)['access_token'],
          scope: JSON.parse(response.body)['scope'],
          auth: self
        )
```

File: lib/arux_app/api/auth.rb [hunk 3]
```diff
@@ -60,6 +60,32 @@ def authorization_url(scope: "public")
...
+      end
+
```
Comment (goronfreeman) [inline lib/arux_app/api/auth.rb:77]:
```suggestion
```

**Review by goronfreeman — COMMENTED** [review]
A couple super small things just from looking at the code. Will test it out tomorrow.

File: lib/arux_app/api/auth.rb (hunk 1 above)
Comment (brgeen) [reply lib/arux_app/api/auth.rb]: Good call on `#tap` works well and improves readability like you mentioned.

File: lib/arux_app/api/auth.rb (hunk 2 above)
Comment (brgeen) [reply lib/arux_app/api/auth.rb]: Nice.

**Review by goronfreeman — APPROVED** [review]

**Pull request merged**